### PR TITLE
SONAR-9884 Load keys of components in short lived branches with issues

### PR DESCRIPTION
**Pull request by janos-ss**

### CODE REVIEW

File: server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDao.java [hunk 1]
```diff
@@ -339,4 +339,8 @@ public void setPrivateForRootComponentUuid(DbSession session, String projectUuid
   public void delete(DbSession session, long componentId) {
     mapper(session).delete(componentId);
   }
+
+  public List<ComponentDto> selectComponentKeysHavingIssuesToMerge(DbSession dbSession, String mergeBranchUuid) {
```
Comment (janos-ss) [inline server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDao.java]: At some point, maybe even in this POC, this should be optimized to not load entire `ComponentDto`, only the two columns in the query result: key and uuid. 
@dbmeneses Let me know if you think this should be done now.

File: server/sonar-db-dao/src/main/resources/org/sonar/db/component/ComponentMapper.xml [hunk 2]
```diff
@@ -646,4 +646,14 @@
...
   </delete>
 
+  <select id="selectComponentKeysHavingIssuesToMerge" resultType="Component">
+    SELECT DISTINCT p.KEE, p.UUID FROM PROJECTS p
+    JOIN ISSUES i
+      ON p.UUID = i.COMPONENT_UUID
+    JOIN PROJECT_BRANCHES b
+      ON i.PROJECT_UUID = b.UUID
+      AND b.MERGE_BRANCH_UUID = #{mergeBranchUuid,jdbcType=VARCHAR}
+      AND i.status != 'OPEN'
```
Comment (janos-ss) [inline server/sonar-db-dao/src/main/resources/org/sonar/db/component/ComponentMapper.xml]: Probably good enough for the POC, but maybe will need tuning.

File: server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDao.java (hunk 1 above)
Comment (janos-ss) [reply server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDao.java]: Nevermind, I'll just do this now, as the performance of this feature is critical for the MMF.

Comment (dbmeneses) [reply server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDao.java]: Yes, I think we should do it now

File: server/sonar-db-dao/src/main/resources/org/sonar/db/component/ComponentMapper.xml (hunk 2 above)
Comment (dbmeneses) [reply server/sonar-db-dao/src/main/resources/org/sonar/db/component/ComponentMapper.xml]: shouldn't this be something like "status == 'CONFIRMED' or (status =='OPEN' AND resolution not null)"? (to be tested with a DB)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ShortBranchComponentsWithIssues.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.commons.lang.StringUtils;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.component.BranchType;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.server.computation.task.projectanalysis.analysis.AnalysisMetadataHolder;
+import org.sonar.server.computation.task.projectanalysis.analysis.Branch;
+
+/**
+ * Cache a map of component key -> uuid in short branches that have issues that are non-open and non-closed.
```
Comment (dbmeneses) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ShortBranchComponentsWithIssues.java]: I think 'non-open and non-closed' is confusing. What about specifying all cases where it applies: status confirmed, or open with resolution FP, won't-fix or fix.

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ShortBranchComponentsWithIssues.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  private Map<String, Set<String>> uuidsByKey;
+
+  public ShortBranchComponentsWithIssues(AnalysisMetadataHolder analysisMetadataHolder, DbClient dbClient) {
+    // TODO maybe can get just the branch here? (unless it's not populated at initialization time)
+    this.analysisMetadataHolder = analysisMetadataHolder;
+    this.dbClient = dbClient;
+  }
+
+  private void loadUuidsByKey() {
+    Branch branch = analysisMetadataHolder.getBranch().get();
```
Comment (dbmeneses) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ShortBranchComponentsWithIssues.java]: I think we can assume that the class is only used when it makes sense (up to the caller!)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ShortBranchComponentsWithIssues.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ */
+public class ShortBranchComponentsWithIssues {
+  private final AnalysisMetadataHolder analysisMetadataHolder;
+  private final DbClient dbClient;
+  private Map<String, Set<String>> uuidsByKey;
+
+  public ShortBranchComponentsWithIssues(AnalysisMetadataHolder analysisMetadataHolder, DbClient dbClient) {
+    // TODO maybe can get just the branch here? (unless it's not populated at initialization time)
+    this.analysisMetadataHolder = analysisMetadataHolder;
```
Comment (dbmeneses) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ShortBranchComponentsWithIssues.java]: IMO we can remove it

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ShortBranchComponentsWithIssues.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    this.dbClient = dbClient;
+  }
+
+  private void loadUuidsByKey() {
+    Branch branch = analysisMetadataHolder.getBranch().get();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ShortBranchComponentsWithIssues.java]: MAJOR Call "Optional#isPresent()" before accessing the value. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3655)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ShortBranchComponentsWithIssues.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+ * Cache a map of component key -> uuid in short branches that have issues that are non-open and non-closed.
+ *
+ */
+public class ShortBranchComponentsWithIssues {
+  private final AnalysisMetadataHolder analysisMetadataHolder;
+  private final DbClient dbClient;
+  private Map<String, Set<String>> uuidsByKey;
+
+  public ShortBranchComponentsWithIssues(AnalysisMetadataHolder analysisMetadataHolder, DbClient dbClient) {
+    // TODO maybe can get just the branch here? (unless it's not populated at initialization time)
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ShortBranchComponentsWithIssues.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by simonbrandhof** [comment]
@janos-ss @dbmeneses https://jira.sonarsource.com/browse/SONAR-9884 is closed, so I suppose that this PR can be closed too ?

**Comment by dbmeneses** [comment]
@simonbrandhof Yes, thanks.

**Pull request closed without merging**